sched/queue/memory/simple.go

```diff
@@ -3,40 +3,134 @@ package memory
 // In-memory Scheduler Queue
 
 import (
-	"fmt"
 	"github.com/scootdev/scoot/sched"
 	"github.com/scootdev/scoot/sched/queue"
+	"strconv"
 	"time"
 )
 
-func NewSimpleQueue() (queue.Queue, chan queue.WorkItem) {
+func NewSimpleQueue(capacity int) queue.Queue {
 	q := &simpleQueue{}
-	q.outCh = make(chan queue.WorkItem, 1)
-	q.inCh = make(chan enqueueReq)
-	go q.loop()
-	return q, q.outCh
+	q.reqCh = make(chan queueReq)
+	q.outCh = make(chan queue.WorkItem)
+
+	st := &simpleQueueState{reqCh: q.reqCh, outCh: q.outCh, capacity: capacity}
+	go st.loop()
+	return q
 }
 
 type simpleQueue struct {
-	nextID int
-	inCh   chan enqueueReq
-	outCh  chan queue.WorkItem
+	// Immutable state, read by many goroutines
+	reqCh chan queueReq
+	outCh chan queue.WorkItem
+}
+
+type simpleQueueState struct {
+	// Mutable state, only read/written by its goroutine
+	reqCh     chan queueReq
+	outCh     chan queue.WorkItem
+	dequeueCh chan struct{}
+	items     []*simpleWorkItem
+	nextID    int
+	capacity  int
+}
+
+func (s *simpleQueueState) loop() {
+	for !s.done() {
+		s.iter()
+
+	}
+	close(s.outCh)
+}
+
+func (s *simpleQueueState) done() bool {
+	// We are done once our inputs are closed and we have no items left to send
+	return s.reqCh == nil && len(s.items) == 0
 }
 
-type simpleWorkItem sched.Job
+func (s *simpleQueueState) iter() {
```

Comment:
All the import stuff is in here and everything else is pretty simple. That's good.

I'm not quite getting how this changes the lack of back pressure when something slow like durability is added. Also, it's a bit hard to reason about sending to a possibly nil outCh and why we set outCh whenever something is enqueued.